### ISSUE TITLE
Dockerfile and third-party libs

### ISSUE THREAD

**Issue by Kirill888** [opendatacube/datacube-core#777]
There has been some discussion on Slack related to `Dockerfile` and various choices that were made with respect to choice of pre-compiled third-party libs. Since Slack message threads disappear let's continue this discussion here.

Some facts about current system in no particular order

- Docker is based of ubuntu:18.04
- Docker uses `ppa:nextgis/ppa` to get more recent `libgdal`
- Docker builds python version of `gdal`
- Docker installs `rasterio` in binary mode, so `rasterio` ships it's own version of `libgdal`
- `rasterio` also ships it's own version of `libcurl`, compiled on Redhat derivative, hence symlink workaround for dealing with `ca-certificates` location
- Tests use `requierments-test.txt` to pin dependencies for 3-d party libs, this is to minimize false positive rate where error is in environment setup and not in our code
- Tests run on travis that uses 16.04 version of ubuntu and also use `nectgis/ppa` to get workable environment.

@woodcockr reported on Slack that he has issues installing `shapely` in the default docker environment.

**Comment by woodcockr** [comment]
Minor correction, the issue is with `cartopy`, `holoviews`, and `shapely` needs a `--no-binary shapely` in order for `cartopy` to not have a Segmentation Fault

**Comment by Kirill888** [comment]
@woodcockr thanks for correction can you please provide more detail as to what did and didn't work and changes you had to make to get the system working for you.

**Comment by Kirill888** [comment]
More facts about current state:

- `libgdal` from `nextgis/ppa` is missing important dependencies so we can not build `rasterio` against it, in particular netcdf and jp2 support is absent.
- `rasterio` doesn't work with `libgdal` 3 and up
- `rasterio` doesn't work with `libgdal` compiled against latest `libproj`, needs `libproj < 6`
- `nextgis/ppa` works for our tests but is not a good fit for sandbox type deployment

**Comment by Kirill888** [comment]
I believe that the best way to deal with all this mess is to compile gdal/geos/proj from source, then build binary wheels against that set of libs:

I have started experimenting with this approach here:

- https://hub.docker.com/r/kkodc/geo-builder
- https://hub.docker.com/r/kkodc/geo-python-base/dockerfile

Basic idea is to have `builder` docker that contains the right clibs, then we use two stage docker to compile needed python wheels in the `builder` then install them into final container, making sure that non-dev versions of correct packages are installed.

**Comment by woodcockr** [comment]
This is both what I did and my current docker image structure since they are related.
Docker images, layer on atop the other:
1. datacube-core:1.7
2. easi-dask: add dask and any additional python libs for the EASI environment -> Holoview and Cartopy added here since these aren't part of datacube-core proper
3. easi-jupyter-singleserver: The jupyter singleserver image, includes jupyter libs and extensions. Built from easi-dask since they need the same python libs on workers.

Modifications to make it work:
datacube-core:1.7 changes
1. Remove the APT installs `python3-*`
2. Add APT installs `proj-bin` `proj-dev` `libgdal-dev` (some were already there)
NOTE: nextgis ppa is still active during this.

easi-dask:
1. Adds `libhdf4-dev libhd5-dev libhdf5-serial-dev netcdf-bin libnetcdf-dev libsnappy-dev`
2. Peculiar hack required for HDF5 headers as follows:
```
RUN ln -s /usr/include/hdf5/serial /usr/include/hdf5/include
ENV HDF5_DIR=/usr/include/hdf5
```
3. Then pip (I've skipped the dask libs, they are the usual ones):
```
pip3 install --upgrade --no-binary shapely \
        Cartopy \
        datashader==0.6.9 \
        fiona \
        flit \
        folium \
        ffmpeg \
        gcsfs==0.2.0 \
        geopandas==0.4.0 \
        geopy \
        geoviews==1.6.2 \
        git+https://github.com/NCAR/intake-thredds \
        graphviz \
        h5netcdf \
        h5py \
        # holoviews (https://github.com/pyviz/holoviews/blob/master/environment.yml) and geoviews (https://github.com/pyviz/geoviews/blob/master/.travis.yml)
        hvplot==0.4.0 \
        intake==0.4.1 \
        intake-xarray==0.3.0 \
        ipyleaflet==0.8.1  \
        ipywidgets==7.4.2 \
        matplotlib \
        netcdf4 \
        pandas \
        rasterio==1.0.18 \
        rasterstats \
        requests \
        ruamel.yaml \
        s3fs==0.2.1 \
        sat-search \
        scikit-image \
        scikit-learn \
        scipy \
        shapely \
        siphon \
        xarray==0.12.0 \
        zarr==2.2.0
```

My next step is to resolve some pinned version differences between the (now) 4 different set of config files in ODC land and the ones pinned above (which are from a holoviews example I was using during testing).

All the iterating on the patch means my branch looks like WIP hell for commits. So far as I can tell the repeatable solution is simply:
1. Drop the APTs for `python3-*` in datacube-core
2. install cartopy and shapely with `--no-binary shapely`, and any missing dependencies from the removal of 1.

I'm still testing, I may well have broken something as a side effect and not come across it yet.

**Comment by alexgleith** [comment]
So, if we were to shift to doing low-level builds, I think we stand on the shoulders of others.

Evan Rault has done some work recently doing GDAL builds, see this as a full Ubuntu example: https://github.com/OSGeo/gdal/blob/master/gdal/docker/ubuntu-full/Dockerfile

Perhaps we should be starting with that image, and installing the ODC and Python dependencies on top of there?

What I would say is that it'll be best if we can maintain the list of Python packages in a requirements.txt file, so that we minimise duplication and try to ensure, again, that we're testing what we're deploying.

**Comment by Kirill888** [comment]
@woodcockr can you include output of:
- `pip3 freeze`
- `gdal-config --version`
- `rio --version`
- `rio --gdal-version`

Anything AWS/s3 related with `rasterio` will be somewhat broken because of `rasterio` version pin. There was a lot changes in that area recently so `rasterio==1.0.18` is problematic as well as old `libgdal`.

**Comment by Kirill888** [comment]
@alexgleith I looked through those and copied a lot of dependency libs from there. Looks like Even is mostly concentrating on 3+ version, understandably so, but due `rasterio` we need `2.4.X`.

**Comment by woodcockr** [comment]
@Kirill888 I doubt the `rasterio==1.0.18` is required for the fix. Just a left over from the debug process since I haven't completed the version rationalisation yet.

Here's the output from the commands:

```
gdal-config --version
2.4.0

rio --version
1.0.18

rio --gdal-version
2.4.0
```
<details><summary>pip3 freeze (click to expand)</summary>

```
affine==2.2.0
amqp==2.2.2
appdirs==1.4.3
asciitree==0.3.3
asn1crypto==0.24.0
astroid==1.6.3
attrs==17.4.0
aws-sam-translator==1.13.0
aws-xray-sdk==2.4.2
backcall==0.1.0
beautifulsoup4==4.8.0
billiard==3.5.0.3
bleach==3.1.0
blosc==1.8.1
bokeh==1.1.0
boto==2.49.0
boto3==1.9.197
botocore==1.12.197
branca==0.3.1
cachetools==2.0.1
Cartopy==0.17.0
celery==4.1.0
certifi==2018.1.18
cf-units==1.2.0
cffi==1.11.5
cfn-lint==0.22.4
cftime==1.0.3.4
chardet==3.0.4
ciso8601==2.1.1
click==6.7
click-plugins==1.0.3
cligj==0.5.0
cloudpickle==0.5.2
colorcet==2.0.1
compliance-checker==3.1.1
coverage==4.5.1
coveralls==1.3.0
cryptography==2.7
cycler==0.10.0
Cython==0.26.1
cytoolz==0.10.0
dask==1.1.0
dask-kubernetes==0.9.0
datacube==0+unknown
datashader==0.6.9
datashape==0.5.2
decorator==4.4.0
defusedxml==0.6.0
dill==0.2.7.1
distributed==1.25.2
docker==4.0.2
docopt==0.6.2
docutils==0.14
ecdsa==0.13.2
entrypoints==0.3
fasteners==0.15
ffmpeg==1.4
Fiona==1.8.6
flit==1.3
folium==0.10.0
future==0.17.1
gcsfs==0.2.0
GDAL==2.4.2
geographiclib==1.49
geopandas==0.4.0
geopy==1.20.0
geoviews==1.6.2
google-auth==1.6.3
google-auth-oauthlib==0.4.0
graphviz==0.11.1
h5netcdf==0.7.4
h5py==2.9.0
HeapDict==1.0.0
holoviews==1.12.3
hvplot==0.4.0
hypothesis==3.55.3
idna==2.6
imageio==2.5.0
intake==0.4.1
intake-thredds==0+untagged.6.g792fdc0
intake-xarray==0.3.0
ipykernel==5.1.1
ipyleaflet==0.8.1
ipython==7.7.0
ipython-genutils==0.2.0
ipywidgets==7.4.2
isodate==0.6.0
isort==4.3.4
jedi==0.14.1
Jinja2==2.10.1
jmespath==0.9.3
joblib==0.13.2
jsondiff==1.1.2
jsonpatch==1.23
jsonpickle==1.2
jsonpointer==2.0
jsonschema==2.6.0
jupyter-client==5.3.1
jupyter-core==4.5.0
kiwisolver==1.1.0
kombu==4.1.0
kubernetes==10.0.0
lark-parser==0.6.7
lazy-object-proxy==1.3.1
llvmlite==0.29.0
locket==0.2.0
lxml==4.2.1
MarkupSafe==1.0
matplotlib==3.1.1
mccabe==0.6.1
mistune==0.8.4
mock==2.0.0
monotonic==1.5
more-itertools==4.1.0
moto==1.3.8
msgpack==0.6.1
msgpack-numpy==0.4.4.3
msgpack-python==0.5.6
multipledispatch==0.6.0
multiprocess==0.70.5
munch==2.3.2
nbconvert==5.5.0
nbformat==4.4.0
netCDF4==1.5.1.2
networkx==2.3
nose==1.3.7
notebook==6.0.0
numba==0.45.0
numcodecs==0.6.3
numpy==1.16.2
oauthlib==3.0.2
objgraph==3.4.0
OWSLib==0.16.0
packaging==19.0
pandas==0.25.0
pandocfilters==1.4.2
param==1.9.1
parso==0.5.1
partd==1.0.0
pathos==0.2.1
pbr==4.0.2
pendulum==1.4.4
pexpect==4.7.0
pickleshare==0.7.5
Pillow==6.1.0
pluggy==0.6.0
pox==0.2.3
ppft==1.6.4.7.1
prometheus-client==0.7.1
prompt-toolkit==2.0.9
protobuf==3.9.0
psutil==5.6.3
psycopg2==2.7.7
ptyprocess==0.6.0
py==1.5.3
pyasn1==0.4.5
pyasn1-modules==0.2.5
pycodestyle==2.4.0
pycparser==2.18
pyct==0.4.6
pygeoif==0.7
Pygments==2.4.2
pygobject==3.26.1
pylint==1.8.4
pyparsing==2.2.0
pyPEG2==2.15.2
pyproj==1.9.5.1
pyshp==2.1.0
pytest==3.5.0
pytest-cov==2.5.1
pytest-httpserver==0.3.0
pytest-timeout==1.2.1
python-apt==1.6.4
python-dateutil==2.7.2
python-jose==3.0.1
python-snappy==0.5.4
pytoml==0.1.21
pytz==2018.4
pytzdata==2018.4
pyviz-comms==0.7.2
PyWavelets==1.0.3
PyYAML==3.12
pyzmq==18.0.2
rasterio==1.0.18
rasterstats==0.13.1
redis==2.10.6
regex==2017.7.28
requests==2.22.0
requests-oauthlib==1.2.0
responses==0.10.6
rsa==4.0
ruamel.yaml==0.16.0
ruamel.yaml.clib==0.1.0
s3fs==0.2.1
s3transfer==0.2.0
sat-search==0.2.1
sat-stac==0.1.3
scikit-image==0.15.0
scikit-learn==0.21.2
scipy==1.3.0
Send2Trash==1.5.0
Shapely==1.6.4.post2
SharedArray==2.0.4
simplejson==3.16.0
singledispatch==3.4.0.3
siphon==0.8.0
six==1.12.0
snuggs==1.4.1
sortedcontainers==2.1.0
soupsieve==1.9.2
SQLAlchemy==1.3.1
tblib==1.4.0
terminado==0.8.2
testpath==0.3.1
toolz==0.9.0
tornado==6.0.3
traitlets==4.3.2
tzlocal==1.5.1
urllib3==1.24.2
vine==1.1.4
wcwidth==0.1.7
webencodings==0.5.1
websocket-client==0.56.0
Werkzeug==0.15.5
widgetsnbextension==3.4.2
wrapt==1.10.11
xarray==0.12.0
xmltodict==0.12.0
zarr==2.2.0
zict==1.0.0
zstandard==0.9.0
```
</details>

**Comment by woodcockr** [comment]
Additional note:
gdal is being restricted in `setup.py` to be `<3` to avoid it being upgraded

**Comment by Kirill888** [comment]
@woodcockr python version of `gdal` really has to match clib version, in your case clib is 2 minor releases behind. GDAL has large surface area and this mismatch probably doesn't matter in the code we use, but who knows. Recommended way is to install same version as clib:

```
pip install GDAL==$(gdal-config --version)
```

So you are only building `shapely` locally and pull the rest with binary dependencies, in particular `rasterio` will bring it's own gdal. I wonder if your pin to 1.0.18 is there to get consistent gdal version between shapely and rasterio libs.

Safest way is to compile all cython libs that depend on gdal against the same libgdal.

You can check if rasterio brings it's own version with this:

```
pip3 show -f rasterio | grep '.libs'
```

**Comment by woodcockr** [comment]
For the information others, working with @Kirill888 who provided me a solid foundation for building the geospatial c-libraries from source followed by python geospatial libraries, we now have a newly minted capability to build out `datacube-core` Docker images ensuring geospatial dependencies are consistent throughout.
There are two new repositories that are required for this to work and then some simplification to the [`datacube-core/Dockerfile`](https://github.com/csiro-easi/datacube-core/blob/develop/Dockerfile) since the new `geo-python` base contains the geospatial libraries already. 
I've trialed this on the CSIRO EASI EC PC setup and I'm going to trail the new images on the CSIRO EASI AWS deployment today as time permits. If this all works I'll tag team with @Kirill888 to place the new repositories into github (I've cloned them into the internal CSIRO git repo at the moment).
We're being a little cautious with this as the downstream ramifications can be rather significant even though it should be the same or better (in my case better since I was having a problem with compatibility with upstream python libraries).

**Comment by alexgleith** [comment]
So, for my $0.02, I think we need should be careful about introducing another 'layer' in our Dockerfile build stack.

Currently it's really clean, and just goes from `ubuntu:18.04` and follows through the same build process as our test suite, so we're (kind of) testing our Docker build when we test our actual dev codebase. I understand we have issues currently, though, so we need to change it up a bit.

If we want a different source for our build, I'd be inclined to still use a `ubuntu`-based image, but would prefer to use one of the [new official osgeo GDAL builds](https://hub.docker.com/r/osgeo/gdal) as a base. The [thinking](http://erouault.blogspot.com/) behind these builds is really neat, and demonstrates great use of Docker paradigms. (Although the lack of a non-3.0 gdal build over there is a showstopper.)

Lastly, if we're going to do our own compilation of libraries, and a multi-stage build, I think doing it all in the ODC Dockerfile would be preferable to having a several layers of Dockerfiles. So again, looking at the above GDAL Dockerfile as a template for what we're doing, just with a bunch of Python library installation after the `FROM ubuntu:18.04 as runner` stage.

**Comment by Kirill888** [comment]
@alexgleith  I looked through gdal dockers quite a bit:

1. They target 3+, there is no 2.4.X on their  docker hub, and `rasterio` demand 2.4.X
2. They are very dev focused, not release focused, (run `rsync` in a separate docker to get `ccache` going and all that, totally justified for development, not so much for release)
3. They are still multi-stage: build in one run in another, otherwise you end up with all your build files and build dependencies.
4. Basing off them would be fine once `rasterio` supports gdal 3+ and proj6, but even then you are forced into a fixed feature set as far as gdal compilation goes.

**Comment by Kirill888** [comment]
@alexgleith image is based on ubuntu:18.04 by the way

**Comment by woodcockr** [comment]
@alexgleith - I agree with @Kirill888. the Osgeo is very similar in approach and has the same level of layering. Our main issue is having sufficient control to finesse the setup so it meets our dependency requirements. It's actually quite simple in the end so I can't see if being a major issue maintaining it.

**Comment by woodcockr** [comment]
So testing in the AWS environment seems to work. I think we can streamline this somewhat to make life easier and do it one hit in the datacube-core dockerfile per part of @alexgleith suggestion.
In any event, its nice to have `holoviews` working. Looks nice in the jlab setup

**Comment by omad** [comment]
@Kirill888 afaict `rasterio` is supposed to work fine with GDAL 3+ and Proj 6. However, they haven't sorted out building wheels against them yet. I haven't tried anything myself yet, but poking around on their repos/issues.

**Comment by Kirill888** [comment]
@omad 

> Rasterio 1.0.x works with Python versions 2.7.x and 3.5.0 through 3.7.x, and GDAL versions 1.11.x through 2.4.x. Official binary packages for Linux and Mac OS X are available on PyPI. Unofficial binary packages for Windows are available through other channels.
> 
> Rasterio 1.0.x is not compatible with GDAL versions 3.0.0 or greater.
> 

on their github readme

**Comment by woodcockr** [comment]
And the 1.0.25 release notes:

> Let's aim for 2019-07-31.
> Note: wheels for PyPI will include GDAL 2.4.2 not GDAL 3.0.
https://github.com/mapbox/rasterio/issues/1734

**Comment by woodcockr** [comment]
but GDAL 3 is part of the CI/CD pipeline so WIP.

**Comment by omad** [comment]
Okay, my bad. Looks like code changes have been merged, but not released yet.

https://github.com/mapbox/rasterio/pull/1729 and https://github.com/mapbox/rasterio/pull/1733

**Comment by Kirill888** [comment]
Most of the work was happening here for now:

https://github.com/Kirill888/geobase

This will be moved into `opendatacube` org at some point, or at least ideas from that repo.

**Comment by Kirill888** [comment]
Ownership was transferred to `opendatacube`

- https://github.com/opendatacube/geobase

Images are pushed to `opendatacube/geobase` from GitHub Actions 
- https://hub.docker.com/repository/docker/opendatacube/geobase/general

There is also a sandbox image, meant to be a replacement for `opendatacube/jupyterhub`, but with working `pyproj|cartopy`
- https://hub.docker.com/repository/docker/opendatacube/sandbox

Experiments with "dev docker for tests and sandbox notebooks" are here:
- https://github.com/opendatacube/geobase/tree/master/odc-dev

@alexgleith

**Comment by alexgleith** [comment]
Hey @woodcockr, how does this look for you? Do you think you'd build images on top of `geobase` or would you prefer to be building on a `datacube-core` image?

I'm inclined for us to be building a `datacube-core` image, in the following way:
1. `opendatacube/datacube-core:latest` is built and tested off the source code in the `develop` branch in this repo
2. we build each release tag into `opendatacube/datacube-core:<release-tag>` and install it from the built pypi packages.

We should add a step of installing the test packages into the latest `datacube-core:latest` build and testing there, before pushing the build without test packages to dockerhub, I think.

I think @Kirill888 feels that we shouldn't have a `datacube-core` image, right? Or is it more that each use-case (like indexing, sandbox, ows, etc.) does their own build, and isn't based on the core image? I'm still not sure how I feel about this... I think having a `core` base image is better, even if it has some issues remaining (i.e., you can't guarantee that the downstream build stages aren't changing the packages).

**Comment by Kirill888** [comment]
@alexgleith the way I see it, current base image solves two problems

1. Sort out python environment with tricky to install dependencies (GDAL, rasterio, pyproj)
2. Provides consistent config for `datacube-core` from environment variables

Problem (1) is solved by geobase in a way that generates more efficient (minimal, no compilers, no test requirements, some of which are quite heavy) and more correct (no binary package clashes) docker images.

Problem (2) is better solved by changing `datacube-core` code to support configuration from environment variable. This avoids problems with file system permissions (config file needs to be written somewhere) and allows for configurations with multiple databases (current config generates just one entry and always overwrites a config file).

**Comment by alexgleith** [comment]
Fair enough.

I guess a question then is do we make an interim `datacube-core` using the new method, so that people with downstream Docker builds can keep working the same, but push them to update to the new suggested method, or do we just leave the old way there, and strongly encourage people to do images different with docs and stuff and deprecate the `datacube-core` image?

**Comment by Kirill888** [comment]
We can make a "compatible-ish"  `datacube-core` image

It will look like this thing:
- https://github.com/opendatacube/geobase/tree/master/odc-index

but with different `requirements.txt`

Config generation is done by this python script:

https://github.com/opendatacube/geobase/blob/master/odc-dev/dc_config_render.py

It should be compatible with the current docker instructions for config but also introduces config from a single environment variable like this:

- `-e DATACUBE_DB_URL="postgresql://username:password@dbhost:5432/database"`

Also that script won't override externally supplied config file.

I think functionality of that config script needs to be copied into `datacube-core` though, then we don't need to have custom entry point at all.

**Comment by Kirill888** [comment]
Problem with docker in general, is that it's basically impossible to define what is "compatible", I'm sure there are down-stream dockers that assume that compilers and various dev libraries are installed in the `datacube-core`, so if we make a more minimal version those will simply break anyway.

**Comment by Kirill888** [comment]
But I also strongly doubt that anyone but us use `opendatacube/datacube-core:latest` as base for anything anyway.

**Comment by woodcockr** [comment]
This looks like it will work for us, similar to the original pattern @Kirill888 setup that CSIRO currently uses but with some of the extra libraries.
The way we are setup in our dev pipeline is will build the full set of images ourselves and tag our own versions accordingly. We do occasionally use a opendatacube supplied image (and often latest ;-) ) if we don't yet have that in our CI/CD pipeline and we wanna play - that is how I did the datacube WMS setup for example. This will work so long as we know the images are tagged to the same version of datacube-core as the one we are using (actually our version may have a variation but that's our problem not odcs).

@tom-butler and I were discussing similar issues with "What is ODC community?" and "What is an organisational thing?" with respect to the datacube k8s on AWS setup. IT impacts which Auto Scaling Groups and EC2 types you use and these are operational issues that can't be specified by ODC community.
What we decided was to include a complete ODC community recipe to deploy all with some reasonable default. This shows an example, then organisations can follow the recipe and do their own thing. Functionally in terraform for this results in the ODC repo being a set of terraform modules that a configured for a common and simple deployment with a few user defined parameters for things like domain name. It avoid the mess of attempting to build modules with every possible variation or embedding organisational choices into the base recipe.

Its more difficult with datacube-core and python dependencies but I think geobase et al is a huge boost to the basic compatibility problem and we can spec the default ODC images so they match with the defaults for ODC EKS deployment with jupyter. (effective but minimal) After that it become organisation specific and you build your own variation on the theme.

Real question is, is there a neat way to provide organisational specific extensions too images at each level?

**Comment by alexgleith** [comment]
> is there a neat way to provide organisational specific extensions too images at each level?

The way I've done things in the past with, say, GeoServer, is use the default until I can't, then use their image as a template to build my own with.

I think what you're saying, around creating an opinionated build, which people can use or extend or simple use as a reference, is a good idea.

**Comment by Kirill888** [comment]
@alexgleith approach you are describing works for "tools", like GeoServer, or Postgress or nginx. But datacube is more of a library than a tool. What one can do with a docker image containing nothing but `datacube` cli tools? Only index/archive datasets to some external db. And even that you won't be able to do unless you install extras like awscli.

Libraries are different, you can't docker-compose `pandas` and `numba` from two different dockers, you can either pick one as base or start from scratch and just cherry pick lines from those dockers. As a library dev best you can do is make it easy to install in a variety of environments your user use.

**Comment by woodcockr** [comment]
For 1.8 release we require

- CSIRO to test the new opendatacube/geobase build facility that GA built
- Add smoke test docker build example works with the Example Jupyterhub setup in datacube-k8s-eks
- Modify the documentation so that the example docker images produced are NOT taken to be production live deployment images. These should be done by an organisations live build system.
- Provide feedback to finalise 

Notes:

- The ODC install runs in a python venv


Open questions:
- Do we build middle wheels docker layer or or not?  use --no-binary in the requirements.txt. It's not entirely necessary
-

**Comment by alexgleith** [comment]
hey @woodcockr, to be clear, we're planning on deprecating the `datacube-core` docker image in 1.8, and providing a barebones Jupyter/Jupyterhub Sandbox-ready image instead.

How do you feel about this approach for Docker?

The alternative is to provide a minimal `datacube-core` image, that does _nothing of value_ except provide a python environment with ODC installed in it...

**Comment by woodcockr** [comment]
Hi @alexgleith Understood, that is what @Kirill888 and I discussed yesterday for ODC 1.8 when we went through in triage for the 1.8 release finalisation (I have a message to send to the ODC SC with an update shortly on that but git got ahead of me whilst I dealt with some other stuff!). 

This will be consistent with the way in which datacube-k8s-eks is operating after 1.8 - ODC community github provides base level examples but any LIVE deployment (images, k8s on AWS, etc) belong to the deploying organisation - so for the docker images you get barebones sandbox, which is actually deployed by the datacube-k8s-eks example but not used by an organisational deployment (that can follow the recipe but isn't an ODC community one).

Oddly enough during the conversation with @Kirill888 it looked to me like this worked better for the datacube-k8s-eks infrastructure as code than it does for docker images (mostly because the use of terraform modules encourages the separation).
